0.2.0a1 (2015-08-13)
####################

*changes since 0.1.7 (2014-05-10)*

Bug Fixes
$$$$$$$$$

New Features
$$$$$$$$$$$$

Changes
$$$$$$$

Security
$$$$$$$$

Other Changes
$$$$$$$$$$$$$

* rework liftover and testing targets [`8df5836c3c8f <https://bitbucket.org/biocommons/uta/commits/8df5836c3c8f>`_]
* update sequence sources config [`352e7f818fab <https://bitbucket.org/biocommons/uta/commits/352e7f818fab>`_]
* standardize, finally, on "NCBI" as origin name; report failed lookups in loading.py [`e8f1e6aa1dcf <https://bitbucket.org/biocommons/uta/commits/e8f1e6aa1dcf>`_]
* removed unused uta.utils [`e720bfbf35bd <https://bitbucket.org/biocommons/uta/commits/e720bfbf35bd>`_]
* create gbff intermediates with N[MR] subsets [`31ab1c081907 <https://bitbucket.org/biocommons/uta/commits/31ab1c081907>`_]
* track records skipped during parsing [`38b4160a1966 <https://bitbucket.org/biocommons/uta/commits/38b4160a1966>`_]
* actually use the specified origin (:-( [`a28af7a9ea19 <https://bitbucket.org/biocommons/uta/commits/a28af7a9ea19>`_]
* loading/Makefile: fix typo bug in gbff glob [`b597a22725b6 <https://bitbucket.org/biocommons/uta/commits/b597a22725b6>`_]
* loading.py: allow missing sequences again -- sigh [`ee9e2adbd6d6 <https://bitbucket.org/biocommons/uta/commits/ee9e2adbd6d6>`_]
* ncbi-parse-gff: ensure that NC transcript structures are used before non-NCs (matters when there's conflict) [`0926cf34e173 <https://bitbucket.org/biocommons/uta/commits/0926cf34e173>`_]
* removed unused imports [`3d7b8737cf0a <https://bitbucket.org/biocommons/uta/commits/3d7b8737cf0a>`_]
* backed out Makefile changes to support GRCh37 (e-t refactoring needed) [`ec3f82bb920f <https://bitbucket.org/biocommons/uta/commits/ec3f82bb920f>`_]
* fixes `#179 <https://bitbucket.org/biocommons/uta/issues/179/>`_ (invitae): added script to pull older geneacs (associated accessions) from eutils [`f0fdaf7abf52 <https://bitbucket.org/biocommons/uta/commits/f0fdaf7abf52>`_]
* check whether mfdb returns None before calling upper() [`9dd9f78cbf5f <https://bitbucket.org/biocommons/uta/commits/9dd9f78cbf5f>`_]
* updated sequence source config [`1ca5497af199 <https://bitbucket.org/biocommons/uta/commits/1ca5497af199>`_]
* updated loading Makefile for dated NCBI downloads (testing needed) [`bc1c2b24d60e <https://bitbucket.org/biocommons/uta/commits/bc1c2b24d60e>`_]
* updated origins [`d37c935ef0eb <https://bitbucket.org/biocommons/uta/commits/d37c935ef0eb>`_]
* reworked divisor/modulus/npartitions logic [`6ca14b365cc9 <https://bitbucket.org/biocommons/uta/commits/6ca14b365cc9>`_]
* ensembl-fetch: use temp dirs; fix bug that causes loop exit on first outer iteration [`7627af13228d <https://bitbucket.org/biocommons/uta/commits/7627af13228d>`_]
* sort genes and split by default to mitigate sporadic api failures [`cfe48d69d4a7 <https://bitbucket.org/biocommons/uta/commits/cfe48d69d4a7>`_]
* added code to investigate bug re: ENSTs with non-distinct sequences (big sigh) [`0546f6f22b55 <https://bitbucket.org/biocommons/uta/commits/0546f6f22b55>`_]
* fixed bug with schema specified in sql/views; minor loading updates [`cca907956058 <https://bitbucket.org/biocommons/uta/commits/cca907956058>`_]
* minor doc updates [`0f719ad37b8b <https://bitbucket.org/biocommons/uta/commits/0f719ad37b8b>`_]
* added indexes to tx_def_summary_mv [`c1cda1cfd4c9 <https://bitbucket.org/biocommons/uta/commits/c1cda1cfd4c9>`_]
* improved handling for multiple versions of ensembl [`cebde5cac68b <https://bitbucket.org/biocommons/uta/commits/cebde5cac68b>`_]
* ensembl-fetch: filter genes by source (hgnc only); fix process splitting; add full GRCh37.p13 aliase set [`7951ad13269f <https://bitbucket.org/biocommons/uta/commits/7951ad13269f>`_]
* dropped ununsed pypi link from README [`09a020fd56b1 <https://bitbucket.org/biocommons/uta/commits/09a020fd56b1>`_]
* Close branch offset. [`757bc7b7c6dc <https://bitbucket.org/biocommons/uta/commits/757bc7b7c6dc>`_]
* Close branch bit. [`5c08a2fef23f <https://bitbucket.org/biocommons/uta/commits/5c08a2fef23f>`_]
* Close branch offset_try2. [`e11f53cc6eea <https://bitbucket.org/biocommons/uta/commits/e11f53cc6eea>`_]
* Close branch dev. [`727b259788f3 <https://bitbucket.org/biocommons/uta/commits/727b259788f3>`_]
* implemented cds_exon_lengths_fp_eq for tx_similarity_v [`590c15c1e38f <https://bitbucket.org/biocommons/uta/commits/590c15c1e38f>`_]
* updated sbin/ensembl-fetch to again support command-line gene names [`a80db96d09bd <https://bitbucket.org/biocommons/uta/commits/a80db96d09bd>`_]
* updated README; added small docs/notes [`0cc79e334157 <https://bitbucket.org/biocommons/uta/commits/0cc79e334157>`_]
* Makefile update for `20150704 <https://bitbucket.org/biocommons/uta/commits/20150704>`_ load [`718065e962f7 <https://bitbucket.org/biocommons/uta/commits/718065e962f7>`_]
* updated load-sequences and added to Makefile (oops) [`6d0ae9b77e6d <https://bitbucket.org/biocommons/uta/commits/6d0ae9b77e6d>`_]
* renamed preferred_accessions to associated_accessions [`38db78fad116 <https://bitbucket.org/biocommons/uta/commits/38db78fad116>`_]
* created doc/figures & relocated files [`d2b14c8c833a <https://bitbucket.org/biocommons/uta/commits/d2b14c8c833a>`_]
* removed uta_20140210/genes.hgnc.gz [`c1cdc8a5f085 <https://bitbucket.org/biocommons/uta/commits/c1cdc8a5f085>`_]
* add more detail during loading & standardize update_interval logic [`3602d602ff48 <https://bitbucket.org/biocommons/uta/commits/3602d602ff48>`_]
* fix bogus chromosome accessions in BIC transcripts during liftover [`52ebdb5c019e <https://bitbucket.org/biocommons/uta/commits/52ebdb5c019e>`_]
* adapt for e-80, and dump results in batches for failure recovery [`51a746c4acbc <https://bitbucket.org/biocommons/uta/commits/51a746c4acbc>`_]
* formatting change only in sql [`2dc653ade771 <https://bitbucket.org/biocommons/uta/commits/2dc653ade771>`_]
* closes `#190 <https://bitbucket.org/biocommons/uta/issues/190/>`_: added hgnc_eq to tx_similarity_v [`c55af9f2313d <https://bitbucket.org/biocommons/uta/commits/c55af9f2313d>`_]
* update serial sequence values during liftover [`75eb4225544a <https://bitbucket.org/biocommons/uta/commits/75eb4225544a>`_]
* removed two stale files [`00444d06bed4 <https://bitbucket.org/biocommons/uta/commits/00444d06bed4>`_]
* bumped schema version (1.0 -> 1.1); in Makefile, get schema name from uta.models dynamically [`87fe6108a5a8 <https://bitbucket.org/biocommons/uta/commits/87fe6108a5a8>`_]
* add progress for seqinfo loading [`0398722d5078 <https://bitbucket.org/biocommons/uta/commits/0398722d5078>`_]
* skip exon sets that are missing transcript definitions; added the-whole-kielbasa loading target [`b0faf4baeae4 <https://bitbucket.org/biocommons/uta/commits/b0faf4baeae4>`_]
* closes `#189 <https://bitbucket.org/biocommons/uta/issues/189/>`_: add _seq_anno_most_recent_v, view with only most recent annotation for a given accession [`9885156aab13 <https://bitbucket.org/biocommons/uta/commits/9885156aab13>`_]
* ensembl: init prior to execution; prevent accidental overwrite; add to mega-test [`39f4ab55f47e <https://bitbucket.org/biocommons/uta/commits/39f4ab55f47e>`_]
* added "mega-test" target in loading Makefile [`715beacb86eb <https://bitbucket.org/biocommons/uta/commits/715beacb86eb>`_]
* fix schema problem with align_exons method in loading.py [`8f235b8c82bb <https://bitbucket.org/biocommons/uta/commits/8f235b8c82bb>`_]
* closes `#188 <https://bitbucket.org/biocommons/uta/issues/188/>`_: implement transcript similarity view (tx_similarity_v) [`40a4e734d606 <https://bitbucket.org/biocommons/uta/commits/40a4e734d606>`_]
* closes `#134 <https://bitbucket.org/biocommons/uta/issues/134/>`_: implement cds clipped fingerprint (in tx_def_summary_v) [`bb07bc46920e <https://bitbucket.org/biocommons/uta/commits/bb07bc46920e>`_]
* bug fix: Makefile bad target name caused preferred-accessions to not load correctly [`70ce152d7ffd <https://bitbucket.org/biocommons/uta/commits/70ce152d7ffd>`_]
* amends `#165 <https://bitbucket.org/biocommons/uta/issues/165/>`_: missing sql/preferred_accession.sql file in commit [`56618feefc18 <https://bitbucket.org/biocommons/uta/commits/56618feefc18>`_]
* closes `#165 <https://bitbucket.org/biocommons/uta/issues/165/>`_: add preferred accessions table and loader (outside of SA) [`b06a874c4cbf <https://bitbucket.org/biocommons/uta/commits/b06a874c4cbf>`_]
* fixed tests to use schema name from uta.models [`cfd6bf12d987 <https://bitbucket.org/biocommons/uta/commits/cfd6bf12d987>`_]
* closes `#185 <https://bitbucket.org/biocommons/uta/issues/185/>`_: use dotted versions in UTA; fix incomplete commit re: origin loading [`186cb3274c1d <https://bitbucket.org/biocommons/uta/commits/186cb3274c1d>`_]
* added bic and ensembl-79 loading targets; updated misc-figures [`4a7cc049d78a <https://bitbucket.org/biocommons/uta/commits/4a7cc049d78a>`_]
* close `#183 <https://bitbucket.org/biocommons/uta/issues/183/>`_: added changelog mechanism [`5c0377bd03f5 <https://bitbucket.org/biocommons/uta/commits/5c0377bd03f5>`_]
* updated URL examples to include schema name [`29a54bee23c8 <https://bitbucket.org/biocommons/uta/commits/29a54bee23c8>`_]
* merged loading overhaul; closes `#177 <https://bitbucket.org/biocommons/uta/issues/177/>`_, `#182 <https://bitbucket.org/biocommons/uta/issues/182/>`_, `#181 <https://bitbucket.org/biocommons/uta/issues/181/>`_, `#180 <https://bitbucket.org/biocommons/uta/issues/180/>`_, `#178 <https://bitbucket.org/biocommons/uta/issues/178/>`_ [`e6b0b6fdd676 <https://bitbucket.org/biocommons/uta/commits/e6b0b6fdd676>`_]
* add load-origin from tsv file; update ensembl Makefile rules [`dc421df5ef19 <https://bitbucket.org/biocommons/uta/commits/dc421df5ef19>`_]
* use str() rather tostring() (deprecated) [`faa2492e212d <https://bitbucket.org/biocommons/uta/commits/faa2492e212d>`_]
* add --prefix argument to ensembl-fetch [`d1ef87aaa67c <https://bitbucket.org/biocommons/uta/commits/d1ef87aaa67c>`_]
* added sbin/pg-dump-schema-rename [`c2fe68fd4371 <https://bitbucket.org/biocommons/uta/commits/c2fe68fd4371>`_]
* merged test.conf into uta_dev@localhost.conf; dropped test.conf [`039faebb0d5a <https://bitbucket.org/biocommons/uta/commits/039faebb0d5a>`_]
* geneacs: write only distinct rows [`e8352bb0dbfb <https://bitbucket.org/biocommons/uta/commits/e8352bb0dbfb>`_]
* explicitly set cds_{start,end}_i to None when not defined in txinfo file being loaded [`e82011d1b02c <https://bitbucket.org/biocommons/uta/commits/e82011d1b02c>`_]
* re-set exec bit on bin/ and sbin/ scripts [`fb727507c17a <https://bitbucket.org/biocommons/uta/commits/fb727507c17a>`_]
* use EMWA for eta; fix sql quoting bugs [`f86ad5d7f601 <https://bitbucket.org/biocommons/uta/commits/f86ad5d7f601>`_]
* reworked to catch NCBI transcript model pathologies; urgent: additional work needed [`5c1e8e3d6112 <https://bitbucket.org/biocommons/uta/commits/5c1e8e3d6112>`_]
* make fasta_directory config a directory list (now fasta_directories) [`004edc2c0d7e <https://bitbucket.org/biocommons/uta/commits/004edc2c0d7e>`_]
* loading doc update [`ac0ecaa16bbd <https://bitbucket.org/biocommons/uta/commits/ac0ecaa16bbd>`_]
* standardized on double-quoted strings [`ff026dbe0e68 <https://bitbucket.org/biocommons/uta/commits/ff026dbe0e68>`_]
* enable loading non-coding transcripts [`63a0fd68b8a2 <https://bitbucket.org/biocommons/uta/commits/63a0fd68b8a2>`_]
* reformat code with autopep8 -r -i . [`ca6bff74f847 <https://bitbucket.org/biocommons/uta/commits/ca6bff74f847>`_]
* enforce FK relationship from exon_set.tx_ac to transcript.ac [`ad5309895b5c <https://bitbucket.org/biocommons/uta/commits/ad5309895b5c>`_]
* misc changes for ensembl-79 loading [`803170ae7aaf <https://bitbucket.org/biocommons/uta/commits/803170ae7aaf>`_]
* update to use bioutils instead of bdi [`769eea9241e4 <https://bitbucket.org/biocommons/uta/commits/769eea9241e4>`_]
* revert to newline-separated fasta_directories [`4fc074482314 <https://bitbucket.org/biocommons/uta/commits/4fc074482314>`_]
* update ensembl-fetch for e79, using GRCh37 mysql instance [`eeb9e4ba73bf <https://bitbucket.org/biocommons/uta/commits/eeb9e4ba73bf>`_]
* minor fix in subset rules [`3e8da2ca39e1 <https://bitbucket.org/biocommons/uta/commits/3e8da2ca39e1>`_]
* make fasta_directory config a directory list (now fasta_directories) [`c2d1630c68f6 <https://bitbucket.org/biocommons/uta/commits/c2d1630c68f6>`_]
* removed stable ncbi-refmismatch directory [`9b6fb1424251 <https://bitbucket.org/biocommons/uta/commits/9b6fb1424251>`_]
* updated subset logic to be indpendent of source data [`410705c53f2d <https://bitbucket.org/biocommons/uta/commits/410705c53f2d>`_]
* updated support for UCSC data [`1e76951479a4 <https://bitbucket.org/biocommons/uta/commits/1e76951479a4>`_]
* added support for data subsets [`912953148609 <https://bitbucket.org/biocommons/uta/commits/912953148609>`_]
* major loading overhaul; see uta/loading/Makefile for current process [`57bd4faa7da3 <https://bitbucket.org/biocommons/uta/commits/57bd4faa7da3>`_]
* removed extraneous views; updated Makefile; corrected bic chromosomes; full update cycle for bic (as demo) now entirely operational [`a018a432d28b <https://bitbucket.org/biocommons/uta/commits/a018a432d28b>`_]
* moved lrg to new set-based loading [`10fcc0526039 <https://bitbucket.org/biocommons/uta/commits/10fcc0526039>`_]
* milestone: new set-based loading experiment works well for bic [`90a51b9adace <https://bitbucket.org/biocommons/uta/commits/90a51b9adace>`_]
* removed old test data [`31d22c28aa69 <https://bitbucket.org/biocommons/uta/commits/31d22c28aa69>`_]
* more loading layout/log changes to support multiple conf options and data sources (not done!) [`f8bf21790d7f <https://bitbucket.org/biocommons/uta/commits/f8bf21790d7f>`_]
* moved older (uta0 and uta1) loading files from loading/main to loading/data [`1c00fdd68379 <https://bitbucket.org/biocommons/uta/commits/1c00fdd68379>`_]
* renamed 'in' directory to 'data' [`c1641e8139d1 <https://bitbucket.org/biocommons/uta/commits/c1641e8139d1>`_]
* added uta subcommand and make rule to update materialized views [`b1f2ef185320 <https://bitbucket.org/biocommons/uta/commits/b1f2ef185320>`_]
* pulled in changes from default [`26878b7f1673 <https://bitbucket.org/biocommons/uta/commits/26878b7f1673>`_]
* updated authors; other minor changes [`d95a22f11cb7 <https://bitbucket.org/biocommons/uta/commits/d95a22f11cb7>`_]
* overhauled a lot (too much); incremental loading works; reworked data set loading [`37f372a3bc14 <https://bitbucket.org/biocommons/uta/commits/37f372a3bc14>`_]
* added uta/formats/ncbitsv.py [`f555fd8e2a15 <https://bitbucket.org/biocommons/uta/commits/f555fd8e2a15>`_]
* updated db_loading.rst [`a069fa5d78c5 <https://bitbucket.org/biocommons/uta/commits/a069fa5d78c5>`_]
* fixed missing uta logo [`2c83138b6ea4 <https://bitbucket.org/biocommons/uta/commits/2c83138b6ea4>`_]
* updated sphinx config, db_loading.rst [`05cc649dbf70 <https://bitbucket.org/biocommons/uta/commits/05cc649dbf70>`_]
* added sbin/ncbi-gff-import [`146d6fe89d73 <https://bitbucket.org/biocommons/uta/commits/146d6fe89d73>`_]
* added sbin/ensembl-fetch-gene-tx-pep to pull ensembl data from biomart [`fbfae1284cd1 <https://bitbucket.org/biocommons/uta/commits/fbfae1284cd1>`_]
* added sbin/ncbi-fetch-gene-nm-np [`7d1bf34ca10c <https://bitbucket.org/biocommons/uta/commits/7d1bf34ca10c>`_]
* added sbin/uta-compare-exonsets [`d6d637d90cee <https://bitbucket.org/biocommons/uta/commits/d6d637d90cee>`_]
* rename sbin/ncbi-fetch to sbin/ncbi-fetch-eutils for clarity [`f67405eb3395 <https://bitbucket.org/biocommons/uta/commits/f67405eb3395>`_]
* Merged in shmumer/testconf-edited-online-with-bitbucket-1424478227206 (pull request #2)
* Close branch shmumer/testconf-edited-online-with-bitbucket-1424478227206 [`2ea2641bbadc <https://bitbucket.org/biocommons/uta/commits/2ea2641bbadc>`_]
* Merged in shmumer/ensemblfetch-edited-online-with-bitbucke-1424902480286 (pull request #3)
* Close branch shmumer/ensemblfetch-edited-online-with-bitbucke-1424902480286 [`7c5a95a0a051 <https://bitbucket.org/biocommons/uta/commits/7c5a95a0a051>`_]
* typo in ensembl-fetch [`14fe9fad33a5 <https://bitbucket.org/biocommons/uta/commits/14fe9fad33a5>`_]
* typo in test.conf [`c05db3c4755c <https://bitbucket.org/biocommons/uta/commits/c05db3c4755c>`_]
* remove sphinx and sphinxcontrib-fulltoc from setup.py to fix build errors on drone.io (akin to hgvs change) [`51bb85271331 <https://bitbucket.org/biocommons/uta/commits/51bb85271331>`_]
* updated urls to bitbucket.org/biocommons [`539f462a87c4 <https://bitbucket.org/biocommons/uta/commits/539f462a87c4>`_]
* added misc/uta-align-test.sh: builds uta-align then tries make test-build in uta, all from scratch [`d2a1b53056be <https://bitbucket.org/biocommons/uta/commits/d2a1b53056be>`_]
* adapt for uta_align updates from Geoff [`59a465eea8f3 <https://bitbucket.org/biocommons/uta/commits/59a465eea8f3>`_]
* loading changes--conf fixes and import changes [`40600baa1460 <https://bitbucket.org/biocommons/uta/commits/40600baa1460>`_]
* fixes `#171 <https://bitbucket.org/biocommons/uta/issues/171/>`_: updated ncbi-fetch to use new(ish) ClientX rather than Client [`d831e02f51fa <https://bitbucket.org/biocommons/uta/commits/d831e02f51fa>`_]
* removed ez_setup.py (seemingly unecessary lately) [`5def70c92d28 <https://bitbucket.org/biocommons/uta/commits/5def70c92d28>`_]
* updated to use the aligner from uta-tools-align (f.k.a. locus-lib-bio) [`a1f492a32f5e <https://bitbucket.org/biocommons/uta/commits/a1f492a32f5e>`_]
* updated loading/README [`e2efe7a8a4ca <https://bitbucket.org/biocommons/uta/commits/e2efe7a8a4ca>`_]
* fixes `#170 <https://bitbucket.org/biocommons/uta/issues/170/>`_: extract duplicated code from UTA and hgvs into a separate package [`e5f776148c5c <https://bitbucket.org/biocommons/uta/commits/e5f776148c5c>`_]
* add bgzip fasta file for test loading [`693d806d3157 <https://bitbucket.org/biocommons/uta/commits/693d806d3157>`_]
* remove loading/quick; added bin/uta-dump-bed; updated README.rst [`d58c22473ff1 <https://bitbucket.org/biocommons/uta/commits/d58c22473ff1>`_]
* fixes `#169 <https://bitbucket.org/biocommons/uta/issues/169/>`_: uta/loading/README now works (at least for the test data) [`cf99fa5957f0 <https://bitbucket.org/biocommons/uta/commits/cf99fa5957f0>`_]
* update repo URL everywhere for move to https://bitbucket.org/uta/uta [`139ac1210a75 <https://bitbucket.org/biocommons/uta/commits/139ac1210a75>`_]
* add mysqldb dependency [`6f67565c1ab6 <https://bitbucket.org/biocommons/uta/commits/6f67565c1ab6>`_]
* README.rst: removed doc link until we have better docs [`750e4b1285dd <https://bitbucket.org/biocommons/uta/commits/750e4b1285dd>`_]
* updated Makefile and .hgignore [`c42a867f2e8a <https://bitbucket.org/biocommons/uta/commits/c42a867f2e8a>`_]
* updated README.rst [`2178ece73478 <https://bitbucket.org/biocommons/uta/commits/2178ece73478>`_]
* moved bermuda table to uta proper and removed from misc/dump/; fixed text that looked like a test in README.rst [`16281ab9aaea <https://bitbucket.org/biocommons/uta/commits/16281ab9aaea>`_]
* added misc/dump/ and update README.rst with instructions for local installation [`030ab8e2c6cc <https://bitbucket.org/biocommons/uta/commits/030ab8e2c6cc>`_]
* updated Makefile, setup.cfg, setup.py [`ece196d08bb2 <https://bitbucket.org/biocommons/uta/commits/ece196d08bb2>`_]
* update Makefile targets for building, testing, uploading [`f55654de80d9 <https://bitbucket.org/biocommons/uta/commits/f55654de80d9>`_]
* `#157 <https://bitbucket.org/biocommons/uta/issues/157/>`_: verified droneio builds now passing [`98c3799de6fe <https://bitbucket.org/biocommons/uta/commits/98c3799de6fe>`_]
